### PR TITLE
Dittodb post

### PR DESCRIPTION
**Pull request by jonkeane**
* [x] I have read the Content Guidelines.
* [x] I have read the Technical Guidelines.
* [x] I used or followed the R Markdown or Markdown template.
* [x] I have followed the Style Guide.
* [x] I created or updated my author metadata with correct folder name.
* [x] I have added relevant tags after browsing existing tags.
* [x] I ran `roblog::ro_lint_md()` on index.md (optional).
* [x] I ran `roblog::ro_check_urls()` on index.md (optional).
* [x] I have added the tags - Software Peer Review, my-packagename, R.
* [x] I have added the package-version YAML tag.
* [x] I have added acknowledgement of the reviewers' work.
* [x] I have added a link to the software peer review thread.

### CODE REVIEW

**Comment by jonkeane** [comment]
My audience for this post is folks who have a decent amount of R experience + some package development so I don't go into too much detail about the basics of object orientation, S4, etc. Though I could add that if the editors think it would be helpful to focus on a different audience. (or add in an explicit "this post is for you if you've developed a package before, even a simple one!" to the introduction)

This might be on the long side, I considered removing the section about `trace`ing to shorten it, but decided to leave it in, at least for the draft. But if something needs to be cut, that was the first thing I was thinking.

**Comment by stefaniebutland** [comment]
Thank you @jonkeane! My colleague @steffilazerte will review.

File: content/blog/2020-08-18-dev-dittodb/index.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
+---
+slug: "dev-dittodb"
+title: Developing dittodb
+package_version: 0.1.1
```
Comment (steffilazerte) [inline content/blog/2020-08-18-dev-dittodb/index.md:4]: Add package version? I know you didn't show any code, but it's helps to pin things down if anything major changes in the future

File: content/blog/2020-08-18-dev-dittodb/index.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+#### Using method dispatch for good
+
+R is famous for having not [one](https://adv-r.hadley.nz/s3.html), not [two](https://adv-r.hadley.nz/s4.html), not [three](http://adv-r.had.co.nz/R5.html) but [four](https://adv-r.hadley.nz/r6.html) object orientation systems.[^10] The DBI package uses [S4](https://adv-r.hadley.nz/s4.html) classes extensively.
+One thing that sets S4 apart from S3 is that S4 allows for formal class inheritance.
+Though a full description of inheritance is out of scope for this post, the basics important for us is that S4 can inherit methods from their parent classes.
+To borrow the inheritance metaphor from [Advanced R](https://adv-r.hadley.nz/s4.html#s4-dispatch):
+
+Say we have a class 🙂 which inherits from class 😶.
+If we write a method `mouth_shape()` for objects of theses classes, we could define one method for each class, so for the 🙂 class which returns the value "smiling" when we call `mouth_shape(🙂)` and then we could define the same method for class 😶 which returns the value `NULL` when we call `mouth_shape(😶)` (since there is no mouth).
```
[the remark this answers is not in the document]
Comment (jonkeane) [reply content/blog/2020-08-18-dev-dittodb/index.md:87]: Heh you can thank Hadley + advanced R for it!

File: content/blog/2020-08-18-dev-dittodb/index.md (hunk 1 above)
Comment (jonkeane) [reply content/blog/2020-08-18-dev-dittodb/index.md:4]: Is there something else that I need to add to make this show up? 0.1.1 is the current version for {dittodb}, I added it here following the technical guides, but I didn't see anything rendered with it on the post itself.

File: content/blog/2020-08-18-dev-dittodb/index.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+For HTTP interfaces this is frequently a chunk of JSON (alongside some HTTP status codes and other information).
+With databases the result is frequently a dataframe (especially for `SEELCT` queries), but can also be lists of information (e.g. `list(completed = 1L)` after checking the status of a query).
+
+The database package (e.g. [RPostgres](https://rpostgres.r-dbi.org) or [RMariaDB](https://rmariadb.r-dbi.org)) is responsible for taking your query, and using a database driver (which almost always use C or C++ based libraries) send the query, get the response from the database, and then transforming that response into an R-native data type (e.g. a dataframe).
+To be successful, when it is operating as a mock database, dittodb must be able to match a specific query to a result that is saved to disk (aka a fixture) and return that instead of executing against an actual database.
+We do this by hashing the query that is sent with `dbSendQuery()` and then we use that hash to look for the fixture file and read that fixture in when there is a call `dbFetch()`.
+
+To make this a concrete example, say you ran the following query while dittodb was mocking the connection[^7]:
```
[the remark this answers is not in the document]
Comment (jonkeane) [reply content/blog/2020-08-18-dev-dittodb/index.md:55]: Absolutely, I've reformatted this section a bit to make the example code more self-contained. I think it's much more approachable now (and is closer to a code chunk on could copy and paste to simply use!). Thank you for the suggestion!

**Comment by jonkeane** [comment]
Thank you so much for the really great comments, they have definitely improved this post.

I've accepted all of the changes — the only thing open at this point is the package version, I _thought_ I had specified it appropriately in the YAML, but I don't see it referenced in the rendered page.

File: content/blog/2020-08-18-dev-dittodb/index.md (hunk 1 above)
Comment (steffilazerte) [reply content/blog/2020-08-18-dev-dittodb/index.md:4]: Oh hmm. I thought 0.1.1 was the place holder. I'm not sure... @maelle, do you know how this works?

**Comment by steffilazerte** [comment]
That's great! I'll do the last pass soon (by early next week, if I don't get to it this week).

File: content/blog/2020-08-18-dev-dittodb/index.md (hunk 1 above)
Comment (jonkeane) [reply content/blog/2020-08-18-dev-dittodb/index.md:4]: If I remember correctly, I changed it from `0.1.0` to `0.1.1`, but regardless `0.1.1` is the actual real version that is released on CRAN. (it was `0.1.0` a bit ago, but was having issues with solars checks so had to bump the version 🙈)

Comment (steffilazerte) [reply content/blog/2020-08-18-dev-dittodb/index.md:4]: I don't even wanna talk about checks on CRAN now :sob: my weathercan package is struggling

Comment (steffilazerte) [reply content/blog/2020-08-18-dev-dittodb/index.md:4]: I think we'll leave this as good enough. Thanks!

File: content/blog/2020-08-18-dev-dittodb/index.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (steffilazerte) [reply content/blog/2020-08-18-dev-dittodb/index.md]: It's funny, I didn't actually mean for this comment to be a suggestion, but I do love the change to footnote :grin:

**Comment by steffilazerte** [comment]
Looks great! Thanks @jonkeane, we don't need anything else from you. We'll just merge on the publication date and you're good to go.

**Comment by jonkeane** [comment]
Thanks for all the super helpful comments!

**Pull request merged**